Add Custom Emojis Support

makotech222 opened the pull request:
This PR adds the custom emoji functionality from https://github.com/LemmyNet/lemmy/issues/2419#issuecomment-1334440588

In the frontend, I've added the following:

* Update post-listing to display a new markdown icon to open emoji picker
 
![image](https://user-images.githubusercontent.com/4389156/207944988-82304efa-dd4b-42f3-a25e-4f76a6745cc6.png)

* Update markdown-textarea to handle emoji-picker
* Update app startup to build emoji list from siteres
* Update Tribute component to include custom emojis
![image](https://user-images.githubusercontent.com/4389156/207945103-69527cf4-5b64-4c59-ae09-b760202d98ee.png)

* Tabify Site Settings page
![image](https://user-images.githubusercontent.com/4389156/207945163-6aef77c2-9e63-46e6-ad73-63a7ddc714fe.png)

* Add Custom Emoji admin tab for add/edit/delete custom emojis. Clicking on emoji in picker will scroll to entry in table.
![image](https://user-images.githubusercontent.com/4389156/207945538-10af5824-9f4e-4019-9313-edb6a00ee378.png)


Deficiencies:
* Bug on emoji-mart: (https://github.com/missive/emoji-mart/issues/756). Functionally, this means the picker can't update when emojis are being changed, unless page is refreshed. Not a show stopper, just annoying.
* Bug on emoji-mart: (https://github.com/missive/emoji-mart/issues/757) Picker blows up if emoji list is empty. Very minor and worked around

Open Questions:
* Better icon for 'save' in the table? Currently using 'edit' icon.
* Is including Image Url column necessary? Allows user to link to external images, but takes up space. I assume most people will just upload images to their server.

Nutomic commented:
I think a good solution would be to split the site settings page into different tabs.

dessalines commented:
Question: is "Category" for emoji-mart? So these all wouldn't just go in a single "custom" category, but potentially many custom categories?

> If we go this route, I think it would be prudent to add api endpoints for PUT/POST/DELETE of the custom emoji entities, rather than use Site.

Sounds good to me.

> Currently, I am just adding a <script> tag to the server's html to load the emoji-mart library. Seeing as how we don't do this for any other library, I imagine we would want to change this. 

Fairly sure you can define that custom library location in the package.json and it will download it to node_modules. If that doesn't work, you can reference the git repo, if the js file is hosted there.

makotech222 commented:
Yeah, Category is something for emoji-mart. You can add emojis to custom categories and it will group them.

makotech222 commented:
Okay with the latest commit, the emoji picker will place a markdown image, with its 'title' attr set to the shortcode. When the markdown renderer encounters an image, it will check if it is a custom emoji, and if so it will add the 'icon-emoji' css class, which limits the size of the image.

dessalines commented:
I'm just getting around to testing this now. There doesn't seem to be a save button for adding the custom emojis.

dessalines commented on src/shared/components/home/admin-settings.tsx:
```diff
@@ -112,21 +114,47 @@ export class AdminSettings extends Component<any, AdminSettingsState> {
+                <button
+                  className={`nav-link btn ${this.state.currentTab == "emojis" && "active"
+                    }`}
+                  onClick={() => { this.handleSwitchTab("emojis"); }}>
```
Use linkEvent here. Check `settings.tsx` for an example.

dessalines commented on src/shared/components/home/admin-settings.tsx:
```diff
@@ -112,21 +114,47 @@ export class AdminSettings extends Component<any, AdminSettingsState> {
+              <li className="nav-item">
+                <button
+                  className={`nav-link btn ${this.state.currentTab == "site" && "active"}`}
+                  onClick={() => { this.handleSwitchTab("site"); }}>
```
Same, use linkEvent.

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+    }
+}
+
+interface EmojiPickerProps {
```
Split these into two files, they are different components.

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+                    <Icon icon="smile" classes="icon-inline" />
+                </button>
+
+                {this.state!.showPicker && (
```
What's the exclamation point do here?

makotech222 replied:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

asserts that object is not null/undefined to the compiler.

makotech222 replied:
noting also, i've fixed most of these anyways

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+        );
+    }
+
+    handleEmojiClick = (e: any) => {
```
Use the conventions with all the other handles, no closures. Strongly type this also.

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+
+    handleEmojiClick = (e: any) => {
+        if (this.props.onEmojiClick)
+            this.props.onEmojiClick(e);
```
Should be able to do `this.props.onEmojiClick?.(...)`

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+
+    handleEmojiClick = (e: any) => {
+        if (this.props.onEmojiClick)
+            this.props.onEmojiClick(e);
```
Same note as above.

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+        );
+    }
+
+    togglePicker = (e) => {
```
Functions, not closures.

makotech222 replied:
I fixed this, but curious why? since it avoids the need to bind(this) on the function in constructor.

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+
+    togglePicker = (e) => {
+        e.preventDefault();
+        this.setState({ showPicker: !this.state!.showPicker });
```
state should be defined already.

makotech222 commented:
Its the icon on each row next to the delete button. Its not a great icon, but couldn't find any better suiting ones.

dessalines commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -328,6 +331,21 @@ export class MarkdownTextArea extends Component<
+  handleEmoji(i: MarkdownTextArea, e: any) {
+    let value = e.native;
+    if (value == null){
+      let lookup = getCustomEmojiLookup()
```
If you've already built the map, just export that value and refer to it, no need to call functions that might accidentally re-initialize things.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+
+interface EmojiFormState {
+    siteRes: GetSiteResponse;
+    custom_emojis: CustomEmojiViewForm[];
```
camelCase

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+            changed: false,
+        }))
+    };
+    private emojiLookup = {};
```
Get rid of this, use maps or functions to do lookups. God I hate javascript.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+        }
+    }
+
+    configurePicker(): any {
```
In the other example, set the picker to {}, why does it need to be configured here but not there?

makotech222 replied:
The admin panel has custom picker options that remove all the built-in emojis and only shows custom emojis

dessalines commented on src/shared/components/common/emoji-picker.tsx:
```diff
@@ -0,0 +1,99 @@
+
+interface EmojiMartProps {
+    onEmojiClick?(val: any): any;
+    pickerOptions: any;
```
Emoji-mart has typescript typings : https://www.npmjs.com/package/@types/emoji-mart

makotech222 replied:
The typings are over 3 years old, unfortunately, so they are basically useless. Ill try to define my own types.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+                    >
+                        <thead className="pointer">
+                            <tr>
+                                <th style="width:70px;">{i18n.t("column_emoji")}</th>
```
Re-do this to bootstrap columns, rather than a table. It'll make it more responsive and usable on mobile, and you won't have to hard-code widths. I need to get rid of every html table in lemmy-ui.

makotech222 replied:
I'm gonna have to push back on this one. Tables are the correct way to display this kind of data. Moving to bootstrap columns would end up wasting a ton of space, as 12 columns isn't specific enough about widths. None of the columns need to be responsive, except for the keywords column, which is set up to expand to fill any remaining width.

I have made a change to optimize it a bit better for mobile by giving the table a min-width so that keywords isn't messed up, but its as usable as its ever gonna be for a small device.

dessalines replied:
Okay, but at the very least remove the hard-coded widths ( those aren't going to be correct for any device but the one you coded them from), and use bootstrap table directives like in `communities.tsx` . As I said though, tables are really terrible for skinny devices like phones, which is why we eventually need to get rid of that, and switch to using responsive design.

Bootstrap has responsive form examples, which is what this is, and we have a good amount of responsive forms in lemmy.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+                            </tr>
+                        </thead>
+                        <tbody>
+                            {this.state!.custom_emojis
```
I think these exclamations can cause crashes right? State should already be initialized, so I'm not sure why the IDE isn't telling you this is redundant.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+                                                htmlFor={index.toString()}
+                                                className="pointer text-muted small font-weight-bold"
+                                            >
+                                                {cv.image_url.length > 0 && <img style="max-width: 24px; max-height: 24px; display: inline-block;" src={cv.image_url} />}
```
Use a style in `main.css`

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+                    <br />
+                    <button
+                        className="btn btn-sm btn-secondary mr-2"
+                        onClick={e => this.handleAddEmojiClick(e)}
```
You have a button to add a block, but no button to save the emojis to lemmy.

dessalines commented:
That icon looks like an edit. Once you turn it from a table into a bootstrap form, it'll probably be better to name the button `Save` for clarity.

makotech222 commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+                                                disabled={cv.id > 0}
+                                                value={cv.shortcode}
+                                                onInput={(s: any) =>
+                                                    this.handleEmojiShortCodeChange(s.target.value, index)
```
With linkEvent, I can't pass any parameters to the function. I can technically embed it in the html element, but that seems so much messier. I've already done same function calls in earlier PRs without linkEvent. From documentation, it just seems like its slightly faster?

dessalines replied:
To pass data to linkEvent, do:

```ts
onClick = {linkEvent(
  {some_data_here},
  this.handleBleh
)}

makotech222 commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,431 @@
+            let currentEmojis = this.state!.custom_emojis;
+            let newEmojiIndex = currentEmojis.findIndex(x => x.shortcode == custom_emoji_view.custom_emoji.shortcode)
+            currentEmojis[newEmojiIndex].id = custom_emoji_view.custom_emoji.id;
+            currentEmojis[newEmojiIndex].changed = false;
```
Not sure what you mean by this. both commentform and postform send websocket requests and implement parseMessage and handle the response. PostForm will notify a parent, via an event, but thats to deal with concerns outside the form. The EmojiForm doesn't need to notify any parents, and handles all its needs internally; nothing in admin-settings needs to know about what happens.

dessalines replied:
Okay I see what you're saying, my bad.

makotech222 commented:
> That icon looks like an edit. Once you turn it from a table into a bootstrap form, it'll probably be better to name the button `Save` for clarity.

Is this okay?

![image](https://user-images.githubusercontent.com/4389156/218280173-32167194-d0ac-4891-b730-6a1477a22fb0.png)

Id prefer a better icon, as its takes up less width, but I couldn't find any in our icon set that worked properly.

makotech222 commented:
Among many fixes today, i added pagination to the admin panel for emojis, to deal with performance issues with large size lists.

dessalines commented on src/shared/utils.ts:
```diff
@@ -74,6 +79,9 @@ export const markdownFieldCharacterLimit = 50000;
 export const relTags = "noopener nofollow";
 
+let customEmojis: EmojiMartCategory[] = [];
+export let customEmojisLookup: Map<string, CustomEmojiView> = new Map<string, CustomEmojiView>();
```
Much better. If we end up getting some weird double calls on this thing tho, we might have to set up a singleton service like was done for UserService and WebsocketService, that could hold some other globals as well.

dessalines commented on src/shared/components/home/admin-settings.tsx:
```diff
@@ -112,21 +114,49 @@ export class AdminSettings extends Component<any, AdminSettingsState> {
-              {this.admins()}
-              {this.bannedUsers()}
-            </div>
+          <div >
```
Weird extra space, wonder how this is passing lint.

dessalines commented on src/shared/components/home/admin-settings.tsx:
```diff
@@ -179,6 +209,11 @@ export class AdminSettings extends Component<any, AdminSettingsState> {
     );
   }
 
+  handleSwitchTab(i: AdminSettings, event: any) {
```
Use the convention for tab switching from `settings.tsx`, line 214 .

dessalines commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -226,6 +228,7 @@ export class MarkdownTextArea extends Component<
             >
               <Icon icon="link" classes="icon-inline" />
             </button>
+            <EmojiPicker onEmojiClick={(e) => this.handleEmoji(this,e)}></EmojiPicker>
```
Use linkEvent

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,456 @@
+    page: number;
+}
+
+class CustomEmojiViewForm {
```
Why is this a class and not an interface?

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,456 @@
+    siteRes: GetSiteResponse;
+    customEmojis: CustomEmojiViewForm[];
+    loading: boolean;
+    page: number;
```
What's going on with page here? I didn't see any paging in the back end.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,456 @@
+            alt_text: x.custom_emoji.alt_text,
+            keywords: x.keywords.map(x => x.keyword).join(" "),
+            changed: false,
+            page: 1 + (Math.floor(index / this.itemsPerPage))
```
Don't hack in paging to the front end, if it needs to be added, it needs to be done properly in the back end as well.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,456 @@
+                            </tr>
+                        </thead>
+                        <tbody>
+                            {this.state.customEmojis.slice((this.state.page - 1) * this.itemsPerPage, ((this.state.page - 1) * this.itemsPerPage) + this.itemsPerPage)
```
Leave complicated front-end paging logic out of this, if we need to later, we can add it in the back end.

makotech222 replied:
Regarding paging: All emojis are already loaded on frontend, no need to pull pages from server. We are paging on admin settings because showing all emojis at once (especially when you get to the thousands) causes page to slow to a crawl and become unusable.

dessalines replied:
If we're really getting to the level of crashing the front end because there are too many emojis, then we really need to take them out of `GetSite` into a paged emoji fetcher. 

Front end hacks like this have to be replicated on all lemmy front ends, which means we're not structuring things correctly. I'm okay with merging this as is, but please add an issue to the Lemmy repo for breaking emojis into its own paged endpoint.

Nutomic replied:
I would consider adding a hardcoded emoji limit in the backend then. If you hit the limit, you just need to remove an existing emoji before adding a new one.

makotech222 replied:
All users need all of the emojis available; when they open the emoji-picker, we need to know what to put in there all at once, not paged. Because of this, its kind of pointless to also have paged endpoints for emojis. I personally consider front-end paging pretty trivial; its mainly necessary once you get to the number of emojis that sites like hexbear have, which ends up with >2000 <tr> elements. Rendering all those at once is unfortunately a weakness of html and will cause page to get unresponsive.

Adding a hard coded limit would be unfortunate, as hexbear would be way over the limit of what is responsive on that page. Seems unnecessary, since the obvious way to work with it is just paginating on the front end. Alternate frontends can go without paginating, as long as they understand that it will get slower as more emojis are added.

dessalines commented on src/shared/components/home/emojis-form.tsx:
```diff
@@ -0,0 +1,456 @@
+        this.setState({ page: page });
+    }
+
+    handleEmojiClick = (e: any) => {
```
No closures, functions only please.

makotech222 commented:
@dessalines Okay I fixed the comments. I updated the table to be more like communities. I do still have the last column with a set width, because the table doesn't really allow two elements in the same column without wrapping it, causing the row width the expand
desktop:
![image](https://user-images.githubusercontent.com/4389156/220424678-01676ac6-cf14-4ab5-a076-121cec48013a.png)
mobile:
![image](https://user-images.githubusercontent.com/4389156/220424751-c1a2fa0a-d45e-4823-a307-052d58c3b055.png)

dessalines reviewed (commented):
Once the back-end gets merged, I'll test this again, and merge it if everything works correctly.

makotech222 commented:
@dessalines Hey, any open requests for changes here? Or just need to test?

dessalines commented:
Okay I deployed the new lemmy-js-client, and merged this fixing a few conflicts.
